### PR TITLE
Initial SceFiber implementation and fixed tls

### PR DESCRIPTION
**Pull request by frangarcj**

### CODE REVIEW

**Comment by isJuhn** [comment]
Latest TLS commits lets R&C progress a bit further, last logged stuff is now some IO stuff with psarc
[vita3k.log](https://github.com/Vita3K/Vita3K/files/4050892/vita3k.log)
Stardew valley also gets slightly further, now complains about bad stack base in TTY and aborts
[vita3k.log](https://github.com/Vita3K/Vita3K/files/4050893/vita3k.log)
nothing new gets rendered in either of them

File: vita3k/cpu/include/cpu/functions.h [hunk 1]
```diff
@@ -39,13 +49,15 @@ uint32_t read_pc(CPUState &state);
 uint32_t read_lr(CPUState &state);
 uint32_t read_fpscr(CPUState &state);
 uint32_t read_cpsr(CPUState &state);
+uint32_t read_TPIDRURO(CPUState &state);
```
Comment (VelocityRa) [inline vita3k/cpu/include/cpu/functions.h]:
```suggestion
uint32_t read_tpidruro(CPUState &state);
```

File: vita3k/cpu/src/cpu.cpp [hunk 2]
```diff
@@ -81,7 +81,7 @@ static void code_hook(uc_engine *uc, uint64_t address, uint32_t size, void *user
 
 static void log_memory_access(uc_engine *uc, const char *type, Address address, int size, int64_t value, MemState &mem) {
     const char *const name = mem_name(address, mem);
-    LOG_TRACE("{}: {} {} bytes, address {} ( {} ), value {}", log_hex((uint64_t)uc), type, size, log_hex(address), name, log_hex(value));
+        LOG_TRACE("{}: {} {} bytes, address {} ( {} ), value {}", log_hex((uint64_t)uc), type, size, log_hex(address), name, log_hex(value));
```
Comment (VelocityRa) [inline vita3k/cpu/src/cpu.cpp]:
```suggestion
    LOG_TRACE("{}: {} {} bytes, address {} ( {} ), value {}", log_hex((uint64_t)uc), type, size, log_hex(address), name, log_hex(value));
```

File: vita3k/cpu/include/cpu/functions.h [hunk 3]
```diff
@@ -54,3 +66,7 @@ bool hit_breakpoint(CPUState &state);
 void trigger_breakpoint(CPUState &state);
 void log_code_add(CPUState &state);
 void log_mem_add(CPUState &state);
+
+CPUContextPtr init_cpu_context();
```
Comment (VelocityRa) [inline vita3k/cpu/include/cpu/functions.h]: not nesessary, include `<memory>` and do `std::make_unique<CPUContext>()`;

File: vita3k/cpu/include/cpu/functions.h [hunk 4]
```diff
@@ -39,13 +49,15 @@ uint32_t read_pc(CPUState &state);
...
 uint32_t read_cpsr(CPUState &state);
+uint32_t read_TPIDRURO(CPUState &state);
 void write_reg(CPUState &state, size_t index, uint32_t value);
 void write_float_reg(CPUState &state, size_t index, float value);
 void write_sp(CPUState &state, uint32_t value);
 void write_pc(CPUState &state, uint32_t value);
 void write_lr(CPUState &state, uint32_t value);
 void write_fpscr(CPUState &state, uint32_t value);
 void write_cpsr(CPUState &state, uint32_t value);
+void write_TPIDRURO(CPUState &state, uint32_t value);
```
Comment (VelocityRa) [inline vita3k/cpu/include/cpu/functions.h]: ditto

File: vita3k/cpu/src/cpu.cpp [hunk 5]
```diff
@@ -408,3 +421,40 @@ void log_mem_add(CPUState &state) {
...
+    }
+
+    ctx.pc = read_pc(state);
+    ctx.lr = read_lr(state);
+    ctx.sp = read_sp(state);
+
+}
+
+void load_context(CPUState &state, CPUContext &ctx) {
+
```
Comment (VelocityRa) [inline vita3k/cpu/src/cpu.cpp]: remove newline

File: vita3k/modules/SceFiber/SceFiber.cpp [hunk 6]
```diff
@@ -25,24 +45,45 @@ EXPORT(int, _sceFiberAttachContextAndSwitch) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, _sceFiberInitializeImpl) {
-    return UNIMPLEMENTED();
+EXPORT(SceInt32, _sceFiberInitializeImpl, emu::SceFiber *fiber, char *name, Ptr<SceFiberEntry> entry, SceUInt32 argOnInitialize, Ptr<void> addrContext, SceSize sizeContext, SceFiberOptParam *params) {
+
```
Comment (VelocityRa) [inline vita3k/modules/SceFiber/SceFiber.cpp]: remove newline

File: vita3k/modules/SceFiber/SceFiber.cpp [hunk 7]
```diff
@@ -61,12 +102,47 @@ EXPORT(int, sceFiberRenameSelf) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceFiberReturnToThread) {
-    return UNIMPLEMENTED();
+EXPORT(SceInt32, sceFiberReturnToThread, SceUInt32 argOnReturn, Ptr<SceUInt32> argOnRun) {
+
```
Comment (VelocityRa) [inline vita3k/modules/SceFiber/SceFiber.cpp]: remove newline

File: vita3k/cpu/include/cpu/functions.h (hunk 3 above)
Comment (frangarcj) [reply vita3k/cpu/include/cpu/functions.h]: I don't understand

File: vita3k/cpu/include/cpu/functions.h (hunk 4 above)
Comment (frangarcj) [reply vita3k/cpu/include/cpu/functions.h]: what?

File: vita3k/kernel/src/thread/thread.cpp [hunk 8]
```diff
@@ -102,6 +102,17 @@ SceUID create_thread(Ptr<const void> entry_point, KernelState &kernel, MemState
         return SCE_KERNEL_ERROR_ERROR;
     }
 
+    thread->cpu_context = init_cpu_context();
```
Comment (VelocityRa) [inline vita3k/kernel/src/thread/thread.cpp]:
```suggestion
    thread->cpu_context = std::make_unique<CPUContext>();
```
`#include <memory>` too

**Pull request merged**